Adopt more smart pointers in DOM code

cdumez opened the pull request:
#### a55f0e9fc70c75494c1898d6c3e87ef8baa9b0df
<pre>
Adopt more smart pointers in DOM code
<a href="https://bugs.webkit.org/show_bug.cgi?id=263632">https://bugs.webkit.org/show_bug.cgi?id=263632</a>

Reviewed by Ryosuke Niwa.

* Source/WTF/wtf/WeakHashCountedSet.h:
(WTF::containerSize):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged):
* Source/WebCore/dom/IdTargetObserver.cpp:
(WebCore::IdTargetObserver::IdTargetObserver):
(WebCore::IdTargetObserver::~IdTargetObserver):
* Source/WebCore/dom/IdTargetObserver.h:
* Source/WebCore/dom/IdTargetObserverRegistry.cpp:
(WebCore::IdTargetObserverRegistry::addObserver):
(WebCore::IdTargetObserverRegistry::removeObserver):
(WebCore::IdTargetObserverRegistry::notifyObserversInternal):
* Source/WebCore/dom/IdTargetObserverRegistry.h:
(WebCore::IdTargetObserverRegistry::notifyObservers):
(WebCore::IdTargetObserverRegistry::IdTargetObserverRegistry): Deleted.
* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueTaskToStartIdlePeriod):
(WebCore::IdleCallbackController::queueTaskToInvokeIdleCallbacks):
* Source/WebCore/dom/ImageOverlay.cpp:
(WebCore::ImageOverlay::hasOverlay):
(WebCore::ImageOverlay::imageOverlayHost):
(WebCore::ImageOverlay::characterRange):
(WebCore::ImageOverlay::isInsideOverlay):
(WebCore::ImageOverlay::isOverlayText):
(WebCore::ImageOverlay::removeOverlaySoonIfNeeded):
(WebCore::ImageOverlay::containerRect):
(WebCore::ImageOverlay::installImageOverlayStyleSheet):
(WebCore::ImageOverlay::updateSubtree):
(WebCore::ImageOverlay::updateWithTextRecognitionResult):
* Source/WebCore/dom/InlineClassicScript.cpp:
(WebCore::InlineClassicScript::create):
* Source/WebCore/dom/InlineStyleSheetOwner.cpp:
(WebCore::InlineStyleSheetOwner::removedFromDocument):
(WebCore::InlineStyleSheetOwner::clearDocumentData):
(WebCore::InlineStyleSheetOwner::clearSheet):
(WebCore::InlineStyleSheetOwner::createSheet):
(WebCore::InlineStyleSheetOwner::sheetLoaded):
(WebCore::InlineStyleSheetOwner::startLoadingDynamicSheet):
* Source/WebCore/dom/KeyboardEvent.cpp:
(WebCore::KeyboardEvent::charCode const):
* Source/WebCore/dom/LiveNodeList.h:
(WebCore::LiveNodeList::protectedDocument const):
* Source/WebCore/dom/LiveNodeListInlines.h:
(WebCore::CachedLiveNodeList&lt;NodeListType&gt;::willValidateIndexCache const):
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::~LoadableNonModuleScriptBase):
(WebCore::LoadableNonModuleScriptBase::hasError const):
(WebCore::LoadableNonModuleScriptBase::notifyFinished):
(WebCore::LoadableNonModuleScriptBase::load):
(WebCore::LoadableClassicScript::execute):
* Source/WebCore/dom/LoadableClassicScript.h:
(WebCore::LoadableNonModuleScriptBase::protectedCachedScript):
* Source/WebCore/dom/LoadableImportMap.cpp:
(WebCore::LoadableImportMap::execute):
* Source/WebCore/dom/LoadableScript.cpp:
(WebCore::LoadableScript::LoadableScript):
(WebCore::LoadableScript::addClient):
(WebCore::LoadableScript::removeClient):
(WebCore::LoadableScript::notifyClientFinished):
* Source/WebCore/dom/LoadableScript.h:
(WebCore::LoadableScript::LoadableScript): Deleted.
* Source/WebCore/dom/LoadableScriptClient.h:
* Source/WebCore/dom/MessageEvent.cpp:
(WebCore::MessageEvent::create):
* Source/WebCore/dom/MessagePort.cpp:
(WebCore::WTF_REQUIRES_LOCK):
(WebCore::MessagePort::create):
(WebCore::MessagePort::disentangle):
(WebCore::MessagePort::messageAvailable):
(WebCore::MessagePort::start):
(WebCore::MessagePort::dispatchMessages):
(WebCore::MessagePort::disentanglePorts):
(WebCore::MessagePort::entangle):
(WebCore::MessagePort::removeEventListener):
(): Deleted.
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::MicrotaskQueue::performMicrotaskCheckpoint):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::initMouseEventQuirk):
(WebCore::MouseEvent::toElement const):
(WebCore::MouseEvent::fromElement const):
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::init):
(WebCore::MouseRelatedEvent::initCoordinates):
(WebCore::MouseRelatedEvent::documentToAbsoluteScaleFactor const):
(WebCore::MouseRelatedEvent::computeRelativePosition):
(WebCore::MouseRelatedEvent::locationInRootViewCoordinates const):
* Source/WebCore/dom/MutationObserver.cpp:
(WebCore::MutationObserver::deliver):
(WebCore::MutationObserver::protectedCallback const):
* Source/WebCore/dom/MutationObserver.h:
* Source/WebCore/dom/MutationObserverRegistration.cpp:
(WebCore::MutationObserverRegistration::MutationObserverRegistration):
(WebCore::MutationObserverRegistration::~MutationObserverRegistration):
(WebCore::MutationObserverRegistration::observedSubtreeNodeWillDetach):
(WebCore::MutationObserverRegistration::shouldReceiveMutationFrom const):
(WebCore::MutationObserverRegistration::isReachableFromOpaqueRoots const):
* Source/WebCore/dom/MutationObserverRegistration.h:
(WebCore::MutationObserverRegistration::protectedObserver):
* Source/WebCore/dom/NamedNodeMap.cpp:
(WebCore::NamedNodeMap::ref):
(WebCore::NamedNodeMap::deref):
(WebCore::NamedNodeMap::getNamedItem const):
(WebCore::NamedNodeMap::isSupportedPropertyName const):
(WebCore::NamedNodeMap::getNamedItemNS const):
(WebCore::NamedNodeMap::removeNamedItem):
(WebCore::NamedNodeMap::element):
(WebCore::NamedNodeMap::protectedElement const):
(WebCore::NamedNodeMap::supportedPropertyNames const):
(WebCore::NamedNodeMap::removeNamedItemNS):
(WebCore::NamedNodeMap::setNamedItem):
(WebCore::NamedNodeMap::item const):
(WebCore::NamedNodeMap::length const):
* Source/WebCore/dom/NamedNodeMap.h:
* Source/WebCore/dom/PendingScript.cpp:
(WebCore::PendingScript::create):
* Source/WebCore/dom/RejectedPromiseTracker.h:

Canonical link: <a href="https://commits.webkit.org/269848@main">https://commits.webkit.org/269848@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21ec5be90d58ffcd2cc22e88fd28b3026a34424c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23787 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1900 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24884 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25935 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21936 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24062 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24287 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24030 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1437 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20563 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26529 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1212 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21473 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27727 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20689 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21685 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21751 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25508 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/23112 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1172 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18865 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/30510 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1172 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6705 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1584 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/30465 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3047 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1488 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6374 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1163,7 +1163,7 @@ inline void StringImpl::deref()
 inline UChar StringImpl::at(unsigned i) const
 {
-    ASSERT_WITH_SECURITY_IMPLICATION(i < m_length);
+    RELEASE_ASSERT(i < m_length);
```
This seems like an unrelated change?

cdumez replied:
Indeed, I must have squashed 2 patches.

cdumez replied:
This is fixed.

brentfulgham commented on Source/WebCore/dom/IdTargetObserverRegistry.h:
```diff
@@ -35,36 +36,30 @@ namespace WebCore {
+    using ObserverSet = HashSet<CheckedRef<IdTargetObserver>>;
+    using IdToObserverSetMap = HashMap<AtomString, std::unique_ptr<ObserverSet>>;
     IdToObserverSetMap m_registry;
     ObserverSet* m_notifyingObserversInSet { nullptr };
```
This is still a gross bare pointer to a HashSet of CheckedRefs. Can this be a CheckedPtr, too?

brentfulgham replied:
Actually, maybe a WeakPtr?

cdumez replied:
We can't, unless we make HashSet subclass CanMakeWeakPtr or CanMakeCheckedPtr. It's unclear we want to make every HashSet bigger.

rniwa replied:
I guess we could wrap in a struct?

cdumez replied:
That's seems like this should be OK. I'll look into doing that.

brentfulgham commented on Source/WebCore/dom/IdTargetObserverRegistry.cpp:
```diff
@@ -69,7 +73,7 @@ void IdTargetObserverRegistry::notifyObserversInternal(const AtomStringImpl& id)
-        m_registry.remove(&id);
+        m_registry.remove(id);
 
     m_notifyingObserversInSet = nullptr;
```
This method makes me wonder why `m_notifyingObserversInSet` isn't just a local variable?

Oh -- I see. We store the pointer as a flag indicating we are doing work, so that we can ASSERT if we attempt to notify observers while we are still notifying observers. But that's only used in DEBUG builds.

I think we should make `m_notifyingObserversInSet` a `bool` that is defined inside a `#ifndef DEBUG` to reflect it's purpose for DEBUG builds, and have a local raw pointer for doing the work in `notifyObserversInternal`.

Or, make the ASSERT in `IdTargetObserverRegistry::notifyObservers` a `RELEASE_ASSERT`.

cdumez replied:
How about the usage of `m_notifyingObserversInSet` in `IdTargetObserverRegistry::removeObserver()` ?

brentfulgham commented on Source/WebCore/dom/ImageOverlay.cpp:
```diff
@@ -117,12 +117,11 @@ bool hasOverlay(const HTMLElement& element)
 {
-    auto host = node.shadowHost();
-    if (!is<HTMLElement>(host))
+    RefPtr host = dynamicDowncast<HTMLElement>(node.shadowHost());
```
Nice improvement!

brentfulgham commented on Source/WebCore/dom/InlineStyleSheetOwner.cpp:
```diff
@@ -96,25 +96,22 @@ void InlineStyleSheetOwner::insertedIntoDocument(Element& element)
         if (scope->hasPendingSheet(element))
             scope->removePendingSheet(element);
         scope->removeStyleSheetCandidateNode(element);
-        scope = nullptr;
```
:-)

brentfulgham commented on Source/WebCore/dom/InlineStyleSheetOwner.cpp:
```diff
@@ -156,36 +153,37 @@ inline bool isValidCSSContentType(const AtomString& type)
-    if (m_sheet) {
-        if (m_sheet->isLoading() && m_styleScope)
-            m_styleScope->removePendingSheet(element);
+    Ref document = element.document();
```
Why not protectedDocument()? I guess that returns a RefPtr?

cdumez replied:
`Ref document = element.document();` is our preferred pattern for local variables. There is no real benefit to calling `protectedDocument()` here? Using `Ref` like I did makes it clear we're dealing with a Ref (a protected document).

brentfulgham commented on Source/WebCore/dom/InlineStyleSheetOwner.cpp:
```diff
@@ -156,36 +153,37 @@ inline bool isValidCSSContentType(const AtomString& type)
-    ASSERT(document.contentSecurityPolicy());
-    const ContentSecurityPolicy& contentSecurityPolicy = *document.contentSecurityPolicy();
-    if (!contentSecurityPolicy.allowInlineStyle(document.url().string(), m_startTextPosition.m_line, text, CheckUnsafeHashes::No, element, element.nonce(), element.isInUserAgentShadowTree())) {
+    ASSERT(document->contentSecurityPolicy());
```
Maybe this should be a RELEASE_ASSERT, since we go on and dereference the potentially nullptr value later on?

cdumez replied:
A null dereference is a safe / very identifiable crash already though.

brentfulgham commented on Source/WebCore/dom/LoadableClassicScript.cpp:
```diff
@@ -64,10 +64,7 @@ bool LoadableNonModuleScriptBase::hasError() const
-        return true;
-
-    return false;
+    return m_cachedScript->errorOccurred();
```
Ha! That's a good simplification.

cdumez committed:
Adopt more smart pointers in DOM code
https://bugs.webkit.org/show_bug.cgi?id=263632

Reviewed by Ryosuke Niwa.

* Source/WTF/wtf/WeakHashCountedSet.h:
(WTF::containerSize):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::attributeChanged):
* Source/WebCore/dom/IdTargetObserver.cpp:
(WebCore::IdTargetObserver::IdTargetObserver):
(WebCore::IdTargetObserver::~IdTargetObserver):
* Source/WebCore/dom/IdTargetObserver.h:
* Source/WebCore/dom/IdTargetObserverRegistry.cpp:
(WebCore::IdTargetObserverRegistry::addObserver):
(WebCore::IdTargetObserverRegistry::removeObserver):
(WebCore::IdTargetObserverRegistry::notifyObserversInternal):
* Source/WebCore/dom/IdTargetObserverRegistry.h:
(WebCore::IdTargetObserverRegistry::notifyObservers):
(WebCore::IdTargetObserverRegistry::IdTargetObserverRegistry): Deleted.
* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueTaskToStartIdlePeriod):
(WebCore::IdleCallbackController::queueTaskToInvokeIdleCallbacks):
* Source/WebCore/dom/ImageOverlay.cpp:
(WebCore::ImageOverlay::hasOverlay):
(WebCore::ImageOverlay::imageOverlayHost):
(WebCore::ImageOverlay::characterRange):
(WebCore::ImageOverlay::isInsideOverlay):
(WebCore::ImageOverlay::isOverlayText):
(WebCore::ImageOverlay::removeOverlaySoonIfNeeded):
(WebCore::ImageOverlay::containerRect):
(WebCore::ImageOverlay::installImageOverlayStyleSheet):
(WebCore::ImageOverlay::updateSubtree):
(WebCore::ImageOverlay::updateWithTextRecognitionResult):
* Source/WebCore/dom/InlineClassicScript.cpp:
(WebCore::InlineClassicScript::create):
* Source/WebCore/dom/InlineStyleSheetOwner.cpp:
(WebCore::InlineStyleSheetOwner::removedFromDocument):
(WebCore::InlineStyleSheetOwner::clearDocumentData):
(WebCore::InlineStyleSheetOwner::clearSheet):
(WebCore::InlineStyleSheetOwner::createSheet):
(WebCore::InlineStyleSheetOwner::sheetLoaded):
(WebCore::InlineStyleSheetOwner::startLoadingDynamicSheet):
* Source/WebCore/dom/KeyboardEvent.cpp:
(WebCore::KeyboardEvent::charCode const):
* Source/WebCore/dom/LiveNodeList.h:
(WebCore::LiveNodeList::protectedDocument const):
* Source/WebCore/dom/LiveNodeListInlines.h:
(WebCore::CachedLiveNodeList<NodeListType>::willValidateIndexCache const):
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::~LoadableNonModuleScriptBase):
(WebCore::LoadableNonModuleScriptBase::hasError const):
(WebCore::LoadableNonModuleScriptBase::notifyFinished):
(WebCore::LoadableNonModuleScriptBase::load):
(WebCore::LoadableClassicScript::execute):
* Source/WebCore/dom/LoadableClassicScript.h:
(WebCore::LoadableNonModuleScriptBase::protectedCachedScript):
* Source/WebCore/dom/LoadableImportMap.cpp:
(WebCore::LoadableImportMap::execute):
* Source/WebCore/dom/LoadableScript.cpp:
(WebCore::LoadableScript::LoadableScript):
(WebCore::LoadableScript::addClient):
(WebCore::LoadableScript::removeClient):
(WebCore::LoadableScript::notifyClientFinished):
* Source/WebCore/dom/LoadableScript.h:
(WebCore::LoadableScript::LoadableScript): Deleted.
* Source/WebCore/dom/LoadableScriptClient.h:
* Source/WebCore/dom/MessageEvent.cpp:
(WebCore::MessageEvent::create):
* Source/WebCore/dom/MessagePort.cpp:
(WebCore::WTF_REQUIRES_LOCK):
(WebCore::MessagePort::create):
(WebCore::MessagePort::disentangle):
(WebCore::MessagePort::messageAvailable):
(WebCore::MessagePort::start):
(WebCore::MessagePort::dispatchMessages):
(WebCore::MessagePort::disentanglePorts):
(WebCore::MessagePort::entangle):
(WebCore::MessagePort::removeEventListener):
(): Deleted.
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::MicrotaskQueue::performMicrotaskCheckpoint):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::initMouseEventQuirk):
(WebCore::MouseEvent::toElement const):
(WebCore::MouseEvent::fromElement const):
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::init):
(WebCore::MouseRelatedEvent::initCoordinates):
(WebCore::MouseRelatedEvent::documentToAbsoluteScaleFactor const):
(WebCore::MouseRelatedEvent::computeRelativePosition):
(WebCore::MouseRelatedEvent::locationInRootViewCoordinates const):
* Source/WebCore/dom/MutationObserver.cpp:
(WebCore::MutationObserver::deliver):
(WebCore::MutationObserver::protectedCallback const):
* Source/WebCore/dom/MutationObserver.h:
* Source/WebCore/dom/MutationObserverRegistration.cpp:
(WebCore::MutationObserverRegistration::MutationObserverRegistration):
(WebCore::MutationObserverRegistration::~MutationObserverRegistration):
(WebCore::MutationObserverRegistration::observedSubtreeNodeWillDetach):
(WebCore::MutationObserverRegistration::shouldReceiveMutationFrom const):
(WebCore::MutationObserverRegistration::isReachableFromOpaqueRoots const):
* Source/WebCore/dom/MutationObserverRegistration.h:
(WebCore::MutationObserverRegistration::protectedObserver):
* Source/WebCore/dom/NamedNodeMap.cpp:
(WebCore::NamedNodeMap::ref):
(WebCore::NamedNodeMap::deref):
(WebCore::NamedNodeMap::getNamedItem const):
(WebCore::NamedNodeMap::isSupportedPropertyName const):
(WebCore::NamedNodeMap::getNamedItemNS const):
(WebCore::NamedNodeMap::removeNamedItem):
(WebCore::NamedNodeMap::element):
(WebCore::NamedNodeMap::protectedElement const):
(WebCore::NamedNodeMap::supportedPropertyNames const):
(WebCore::NamedNodeMap::removeNamedItemNS):
(WebCore::NamedNodeMap::setNamedItem):
(WebCore::NamedNodeMap::item const):
(WebCore::NamedNodeMap::length const):
* Source/WebCore/dom/NamedNodeMap.h:
* Source/WebCore/dom/PendingScript.cpp:
(WebCore::PendingScript::create):
* Source/WebCore/dom/RejectedPromiseTracker.h:

Canonical link: https://commits.webkit.org/269848@main
